src/Http/WebUtilities/test/MultipartPipeReaderTests.cs

```diff
@@ -0,0 +1,376 @@
+        }
+
+        [Fact]
+        public async Task MultipartPipeReader_ReadTwoPartBodyWithUnicodeFileName_Success()
+        {
+            var pipeReader = MakeReader(TwoPartBodyWithUnicodeFileName);
+            var reader = new MultipartPipeReader(Boundary, pipeReader);
+
+            var section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Single(section!.Headers);
+            Assert.Equal("form-data; name=\"text\"", section!.Headers!["Content-Disposition"][0]);
+
+            var result = await section!.ReadAsStringAsync();
+            Assert.Equal("text default", result);
+
+            section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Equal(2, section!.Headers?.Count);
+            Assert.Equal("form-data; name=\"file1\"; filename=\"a色.txt\"", section!.Headers!["Content-Disposition"][0]);
+            Assert.Equal("text/plain", section!.Headers!["Content-Type"][0]);
+
+            result = await section!.ReadAsStringAsync();
+            Assert.Equal("Content of a.txt.\r\n", result);
+
+            Assert.Null(await reader.ReadNextSectionAsync());
+        }
+
+        [Fact]
+        public async Task MultipartPipeReader_ThreePartBody_Success()
+        {
+            var pipeReader = MakeReader(ThreePartBody);
+            var reader = new MultipartPipeReader(Boundary, pipeReader);
+
+            var section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Single(section!.Headers);
+            Assert.Equal("form-data; name=\"text\"", section!.Headers!["Content-Disposition"][0]);
+            var result = await section!.ReadAsStringAsync();
+            Assert.Equal("text default", result);
+
+            section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Equal(2, section!.Headers?.Count);
+            Assert.Equal("form-data; name=\"file1\"; filename=\"a.txt\"", section!.Headers!["Content-Disposition"][0]);
+            Assert.Equal("text/plain", section!.Headers!["Content-Type"][0]);
+            result = await section!.ReadAsStringAsync();
+            Assert.Equal("Content of a.txt.\r\n", result) ;
+
+            section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Equal(2, section!.Headers?.Count);
+            Assert.Equal("form-data; name=\"file2\"; filename=\"a.html\"", section!.Headers!["Content-Disposition"][0]);
+            Assert.Equal("text/html", section!.Headers!["Content-Type"][0]);
+            result = await section!.ReadAsStringAsync();
+            Assert.Equal("<!DOCTYPE html><title>Content of a.html.</title>\r\n", result);
+
+            Assert.Null(await reader.ReadNextSectionAsync());
+        }
+
+        [Fact]
+        public async Task MultipartPipeReader_TwoPartBodyIncompleteBuffer_OneSectionReadsSuccessfullySecondSectionThrows()
```

Comment:
I fixed the test to pass, but I am not sure if it's a functional requirement or not - Adding a subsequent AdvanceTo and ReadAsync will throw, but I am not sure if that's what we want. 
Basically, it's conflicting with a comment inside the MultipartSectionPipeReader  _We want them to consume and advance all of the data before we try to drain the metadata / boundary._

Reply:
Looking at the original now, I am not sure I would say that the behavior had changed.
If we alter the original test and add another read to the section.Body it would fail.
If we alter this test to read the third section instead of continuing to read the rest of the second section it would also fail.

I think that this behavior was changed with my original implementation of the PipeReader, and the test just wasn't updated.

Anyways, we might want to consider altering the test to reflect the actual scenario - basically the behavior in both the Stream implementation and the Pipe Implementation is that for an incomplete Section Boundry - the second section will throw before reaching the end - (Pipe - IsCompleted == true, Stream - Read returns 0) - and trying to read the third section will throw.  I would argue that in reality we are more likely to try and read the second section to it's end before attempting to read the third section, especially since we will probably copy each section to it's own Pipe.